File: x/wasmx/keeper/testdata/js/blockchain.js
import * as wasmx from 'wasmx';

export function instantiate(dataObj) {
    return store(...dataObj);
}

export function main(dataObj) {
    // console.log("-***--main-dataObj-", dataObj)
    console.log("-***--main-dataObj-");
    console.log("-***--main-dataObj-keys-", Object.keys(dataObj));
    if (dataObj.getEnv) return getEnv();
    if (dataObj.getCallData) return getCallData();
    if (dataObj.store) return store(...dataObj.store);
    if (dataObj.load) return load();
    if (dataObj.getBlockHash) return getBlockHash(...dataObj.getBlockHash);
    if (dataObj.getAccount) return getAccount(...dataObj.getAccount);
    if (dataObj.getBalance) return getBalance(...dataObj.getBalance);
    if (dataObj.keccak256) return keccak256(...dataObj.keccak256);
    if (dataObj.createAccount) return createAccount(...dataObj.createAccount);
    if (dataObj.createAccount2) return createAccount2(...dataObj.createAccount2);
    throw new Error("no valid function");
}

function getEnv() {
    console.log("-*****-getEnv");
    const envbuf = wasmx.getEnv();
    console.log("--envbuf", envbuf);
    const env = JSON.parse(arrayBufferToString(envbuf));
    console.log("--env", env);
    return stringToArrayBuffer(JSON.parse(env));
}

function getCallData() {
    const calldbuf = wasmx.getCallData();
    console.log("--calldbuf", calldbuf);
    const data = JSON.parse(arrayBufferToString(envbuf));
    console.log("--data", data);
    return stringToArrayBuffer(JSON.parse(data));
}

function getBlockHash(address) {

}

function getAccount(address) {

}

function getBalance(address) {
    const bz = wasmx.bech32StringToBytes(address);
    const balance = wasmx.getBalance(bz);
    return balance;
}

function keccak256(data) {

}

function createAccount() {

}

function createAccount2() {

}

function store(key_, value_) {
    console.log("-store-dataObj-key_-", key_)
    console.log("-store-dataObj-value_-", value_)
    const key = new Uint8Array(key_);
    const value = new Uint8Array(value_);
    console.log("-store-dataObj-key-", key)
    console.log("-store-dataObj-value-", value)
    wasmx.storageStore(key.buffer, value.buffer);
}

function load(key_) {
    const key = new Uint8Array(key_);
    return wasmx.storageLoad(key);
}

// utils

function stringToArrayBuffer(inputString) {
    const bytes = new Uint8Array(inputString.length);
    for (let i = 0; i < inputString.length; i++) {
        bytes[i] = inputString.charCodeAt(i) & 0xFF;
    }
    return bytes.buffer;
}

function arrayBufferToString(arrayBuffer) {
    const bytes = new Uint8Array(arrayBuffer);
    let result = "";
    for (let i = 0; i < bytes.length; i++) {
        result += String.fromCharCode(bytes[i]);
    }
    return result;
}

const hexStringToArrayBuffer = hexString => new Uint8Array(hexString.match(/.{1,2}/g).map(byte => parseInt(byte, 16))).buffer;

File: tests/testdata/python/blockchain.py
import json
import wasmx

def instantiate(dataObj):
    return store(*dataObj)

def main(dataObj):
    if "justError" in dataObj:
        return justError()
    if "getEnv" in dataObj:
        return getEnv_()
    if "getCallData" in dataObj:
        return getCallData_()
    if "store" in dataObj:
        return store(*dataObj["store"])
    if "load" in dataObj:
        return load()
    if "getBlockHash" in dataObj:
        return getBlockHash_(*dataObj["getBlockHash"])
    if "getAccount" in dataObj:
        return getAccount_(*dataObj["getAccount"])
    if "getBalance" in dataObj:
        return getBalance_(*dataObj["getBalance"])
    if "keccak256" in dataObj:
        return keccak256_(*dataObj["keccak256"])
    if "instantiateAccount" in dataObj:
        return instantiateAccount(*dataObj["instantiateAccount"])
    if "instantiateAccount2" in dataObj:
        return instantiateAccount2(*dataObj["instantiateAccount2"])
    wasmx.set_exit_code(1, 'Invalid function')

def justError():
    wasmx.set_exit_code(1, 'just error')

def getEnv_():
    envbuf = wasmx.get_env()
    envstr = arrayBufferToString(envbuf)
    env = json.loads(envstr)
    return stringToArrayBuffer(json.dumps(env))

def getCallData_():
    calldbuf = wasmx.get_calldata()
    data = json.loads(arrayBufferToString(calldbuf))
    return stringToArrayBuffer(json.dumps(data))

def getBlockHash_(blockNumber):
    return wasmx.get_blockhash(blockNumber)

def getAccount_(address):
    addrbuf = wasmx.bech32_string_to_bytes(address)
    accountbuf = wasmx.get_account(addrbuf)
    account = json.loads(arrayBufferToString(accountbuf))
    return stringToArrayBuffer(json.dumps(account))

def getBalance_(address):
    bz = wasmx.bech32_string_to_bytes(address)
    return wasmx.get_balance(bz)

def keccak256_(data):
    databz = stringToArrayBuffer(data)
    return wasmx.keccak256(databz)

def instantiateAccount(codeId, initMsg, balance):
    msgbuf = hexStringToArrayBuffer(initMsg)
    balancebuf = hexStringToArrayBuffer(balance)
    return wasmx.instantiate(codeId, msgbuf, balancebuf)

def instantiateAccount2(codeId, salt, initMsg, balance):
    return wasmx.instantiate2(codeId, hexStringToArrayBuffer(salt), hexStringToArrayBuffer(initMsg), hexStringToArrayBuffer(balance))

def store(key_, value_):
    key = stringToArrayBuffer(key_)
    value = stringToArrayBuffer(value_)
    wasmx.storage_store(key, value)

def load(key_):
    key = stringToArrayBuffer(key_)
    return wasmx.storage_load(key)

# utils

def stringToArrayBuffer(inputString):
    return inputString.encode()

def arrayBufferToString(arrayBuffer):
    return arrayBuffer.decode()

def hexStringToArrayBuffer(hex_string):
    return bytes.fromhex(hex_string)
